enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,989 @@
+media containing an upgrade bundle) has been plugged. To control that a new
+`spec.desiredUpdate.bundle.monitor` field will be added to the `ClusterVersion`
+object. The default value will be `false` so that this monitoring will be
+disabled by default. When set to `true` the cluster version operator will create
+a new `bundle-monitor` daemon set that will perform the actual monitoring. When
+any of the pods in this daemon set detects a device containing a valid upgrade
+bundle it will update the status of the `ClusterVersion` object indicating that
+the bundle is available, via a new `status.desired.bundle.file` field.
+
+For example, if the administrator of the cluster wants to enable automatic
+detection of upgrade bundles she will add this to the `ClusterVersion` object:
+
+```yaml
+apiVersion: config.openshift.io/v1
+kind: ClusterVersion
+metadata:
+  name: version
+spec:
+  desiredUpdate:
+    bundle:
+      enabled: true
+      monitor: true
+```
+
+For situations where it isn't desirable to monitor device events, or it isn't
+possible to plug the USB stick or any other kind of media, we will also add a
+new `spec.desiredUpdate.bundle.file` field to explicitly indicate the location
+of the bundle. When this is used the value will be directly copied to the
+`status.desired.bundle.file` field, without creating the `bundle-monitor` daemon
+set.
+
+If the administrator wants to disable automatic detection of the upgrade bundle,
+and wants to manually copy the file to one of the cluster nodes she will do
+something like this:
+
+```yaml
+spec:
+  desiredUpdate:
+    bundle:
+      enabled: true
+      file: /root/upgrade-4.13.7-x86_64.tar
+```
+
+If the administrator wants to disable automatic detection, but wants to use a
+USB stick anyhow he will do something like this:
+
+```yaml
+spec:
+  desiredUpdate:
+    bundle:
+      enabled: true
+      file: /dev/sdb
+```
+
+When the `status.desired.bundle.file` field is populated the cluster version
+operator will start to replicate the bundle to the rest of the nodes of the
+cluster. To do so it will first disable auto-scaling to ensure that no new
+nodes are added to the cluster while this process is in progress. Then it will
+start a new `bundle-server` daemon set. Each of the pods in this daemon set
+will check if the file specified in the `status.desired.bundle.file` field
+exists and contains a valid bundle. If it does then it will serve it via HTTP
+for the other nodes of the cluster. If it doesn't exist then it will do
+nothing.
+
+Simultaneously with the `bundle-server` daemon set the cluster version operator
+will also start a new `bundle-extractor` batch job in each node of the cluster.
+Each pod in these jobs will try to read the bundle from the location specified
+in `status.desired.bundle.file` field. If that file doesn't exist it will try
+to download it from the HTTP server of one of the pods of the `bundle-server`
```

Comment:
Downloading a single 32G file over HTTP is going to be really fragile...I think we should really be moving more in the direction of transient registry servers and peer-to-peer fetches.

Reply:
It is 16 GiB if we use a dump of a docker/distribution registry server as we suggest. But I agree it is fragile. Actually initially I thought about using bittorrent, but had some difficulties to set it up in the prototype. If that worked it would be better for reliability and performance for large clusters. There won't be much difference for small clusters. Is that peer-to-peer fetches feature a reality? Is someone working on it already? If it was it would certainly help here.